File: python/cucumber_expressions/generated_expression.py
class GeneratedExpression:
    def __init__(self, expression_template: str, parameter_types):
        self.expression_template = expression_template
        self.parameter_types = parameter_types

    @property
    def source(self):
        return self.expression_template % tuple([p.name for p in self.parameter_types])

    @property
    def parameter_names(self):
        usage_by_type_name = {}
        return [
            self.get_parameter_name(t.name, usage_by_type_name)
            for t in self.parameter_types
        ]

    @staticmethod
    def get_parameter_name(type_name, usage_by_type_name):
        count = usage_by_type_name.get(type_name) or 0
        count = count + 1
        usage_by_type_name[type_name] = count
        return type_name if count == 1 else f"{type_name}{count}"

File: javascript/src/GeneratedExpression.ts
import ParameterType from './ParameterType.js'
import { ParameterInfo } from './types.js'

export default class GeneratedExpression {
  constructor(
    private readonly expressionTemplate: string,
    public readonly parameterTypes: readonly ParameterType<unknown>[]
  ) {}

  get source() {
    return format(this.expressionTemplate, ...this.parameterTypes.map((t) => t.name || ''))
  }

  /**
   * Returns an array of parameter names to use in generated function/method signatures
   *
   * @returns {ReadonlyArray.<String>}
   */
  get parameterNames(): readonly string[] {
    return this.parameterInfos.map((i) => `${i.name}${i.count === 1 ? '' : i.count.toString()}`)
  }

  /**
   * Returns an array of ParameterInfo to use in generated function/method signatures
   */
  get parameterInfos(): readonly ParameterInfo[] {
    const usageByTypeName: { [key: string]: number } = {}
    return this.parameterTypes.map((t) => getParameterInfo(t, usageByTypeName))
  }
}

function getParameterInfo(
  parameterType: ParameterType<unknown>,
  usageByName: { [key: string]: number }
): ParameterInfo {
  const name = parameterType.name || ''
  let counter = usageByName[name]
  counter = counter ? counter + 1 : 1
  usageByName[name] = counter
  let type: string | null
  if (parameterType.type) {
    if (typeof parameterType.type === 'string') {
      type = parameterType.type
    } else if ('name' in parameterType.type) {
      type = parameterType.type.name
    } else {
      type = null
    }
  } else {
    type = null
  }
  return {
    type,
    name,
    count: counter,
  }
}

function format(pattern: string, ...args: readonly string[]): string {
  return pattern.replace(/{(\d+)}/g, (match, number) => args[number])
}
